### PR TITLE
[skyrl][tinker] Use VLLMRenderer in SkyRL train backend

### LINKED ISSUE #1200: [skyrl][RFC] VLM Training in SkyRL with the SkyRL train backend
# VLM Training in SkyRL

This document outlines the changes required for SkyRL/skyrl to support vision inputs, similar to the official tinker implementation. The proposed changes are only applicable to the pytorch / skyrl-train backend. The jax backend is out of scope.

_Test Case_: The supervised vision language training example ([here](https://github.com/thinking-machines-lab/tinker-cookbook/tree/main/tinker_cookbook/recipes/vlm_classifier)) should run and converge. The VLM SFT example should work out of the box. The task is to classify CalTech101, a very standard image object recognition benchmark (expected accuracy >90%, especially post-SFT).

An example implementation is on this [branch](https://github.com/nithinvc/SkyRL/pull/2/changes).

## Control Flow Overview
The largest change is passing the list of `ModelInputChunks` to the SkyRL-train backend. This is to reconcile the fact that pytorch / huggingface expects sequences with image token padding while VLLM inserts image padding within the engine. The proposed workaround is to pass the `ImageChunk`s to SkyRL-train which decides, based on the method call, whether to add image pad tokens or not.

Image

## Tokenization Responsibilities
The client handles tokenization of text and encoding images into bytes. The server then has two code paths: 
1) Sampling through vllm and 
2) Training with huggingface / fsdp2.

VLLM: skyrl-train will be responsible for converting the bytes64 into a pillow image. This will then be passed through a VLLM [renderer](https://docs.vllm.ai/en/v0.10.2/api/vllm/entrypoints/renderer.html).
 
HF / FSDP2 Training: Because the datum is a list of encoded chunks which need to get concatenated, the skyrl-train backend should:
1. Patchify the image
2. Compute the expected number of tokens (validate if provided by client)
3. Concatenate input chunks to create the model input

## Tasks Breakdown
There’s a lot of moving pieces to get to a full VLM SFT implementation. Instead, here are a set of incremental changes which should preserve backwards compatibility and interoperability with the Jax backend.

**1. Inference Engine Piping**
Changes localized around the `sample` workflow. We will only support the new inference stack (`inference_engine_client_http_endpoint.py`).
- [x] Inference engine client call for `/v1/chat/completion/render` + inclusion of multi-modal data.
- [x] Implement the sample API which performs the generation by calling `/inference/v1/generate`.
- [ ] [VLLM Upstream] Implement multi-modal features in `inference/v1/generate`. [Here](https://github.com/vllm-project/vllm/blob/main/vllm/entrypoints/serve/disagg/protocol.py#L33) - renderer refactor is completed, this is primarily a piping issue.
- [ ] InferenceEngineClient calls the VLLM renderer endpoint to get a `GenerateRequest`, which is expected by the `v1/generate` api.
- [x] Verify existing RL / SFT training loops work with new inference server stack. 

_Compatibility_: Existing sample calls should not be affected. Multi-modal data piping defaults to None.
_Tests_: a) Integration test with a VL model. Pass a red square and ask it to answer what color it is. b) CI test verifying our pre-processing yields the same result as VLLM, along with a log probabilities comparison test.

**2. Training Integration (FSDP2 only)**
- [x] Pass multi-modal inputs through workers (Policy worker and Ref worker). In worker.py update `_forward_backward_micro` to pull out `multi_modal_inputs` and pass to `model_wrapper.py`.
- [x] Update `model_wrapper.py` forward call signature to accept multi-modal inputs. Pass to FSDP2 wrapped huggingface model.

_Compatibility_: New prompt ID construction is only used when image chunks are present. Otherwise, multi-modal data defaults to None.
_Tests_: Unit tests for `image_chunk_to_image_tensor` and `chunks_to_input_ids`. Update `Experience` and `TrainingInput` tests.

**3. SkyRL-train chunk processing**
- [x] Pass chunks to SkyRL-train backend.
    - [x] `ModelInputChunk` aggregation and pass through for `prepare_sample_batch` 
    - [x] `ModelInputChunk` aggregation and pass through for `prepare_model_pass_batch`
- [x] Update the jax backend to accept chunks, but no-op with them.
- [x] In SkyRL-train, when chunks are provided, construct text only inputs (no vision yet).
- [x] Remove concatenation within the engine, and move it to the backend. Comparing between git commits should yield the same result.

_Compatibility_: Existing codepaths expecting a list[int] for prompt-ids should still work.
_Tests_: Multiple encoded text chunks should yield the same result as a single encoded text chunk. Similarly multiple encoded text chunks should be the same as not passing `ModelChunks` to the backend.

**4. Core tinker types expansion**
- [x] Extend the definition of `ModelInputChunk` beyond text (generalized `InputChunk`). [Tinker SDK](https://github.com/thinking-machines-lab/tinker/blob/main/src/tinker/types/image_chunk.py) 
- [x] Add `ImageChunk` definition from SDK.
- [x] [New `renderer.py`] Create a renderer which uses VLLM to process vision inputs and create an appropriate `token_ids` and `pixel_values` for the training backend.

_Compatibility_: No new control paths for text. Vision language training.
_Tests_: Test initialization and chunk validation logic. Correctness test via CalTech101 tinker cookbook.

### PR DESCRIPTION
**Pull request by nithinvc**
## Summary

Integrates the VLLMRenderer (landed in #1464) into the SkyRL train backend so that VLM training batches include image placeholder tokens and decoded vision tensors (`pixel_values`, `image_grid_thw`).

- When using new inference (`_SKYRL_USE_NEW_INFERENCE`), `_to_training_batch` lazily creates a `VLLMRenderer` and renders all `ModelInput`s through it. 
- Extracts `pixel_values` and `image_grid_thw` from rendered outputs and adds them to the `TrainingInputBatch` as `TensorList` entries (one tensor per batch element, since patch counts vary per image).
- Extends `_pad_batch` to handle `TensorList` fields by cycling and cloning entries, matching the existing padding strategy for regular tensors.
- Reorders `forward_backward` and `forward` to call `_to_training_batch` before `_sleep_inference_engines`, since the renderer needs the inference servers need to be initialized. Note that this does not wake the KV cache or model GPU memory since that is explicitly done in `save_weights_for_sampler` via the dispatcher. 

## E2E Tinker VLM Classifier Curves
With #1484 , we can now run tinker vision language recipes against the SkyRL. Merging both closes #1200 

Example:
```bash
 _SKYRL_USE_NEW_INFERENCE=1 uv run --extra tinker --extra fsdp -m skyrl.tinker.api \
    --base-model "Qwen/Qwen3-VL-8B-Instruct" \
    --backend fsdp \
    --backend-config '{"trainer.placement.policy_num_gpus_per_node": 8, "generator.inference_engine.num_engines": 8, "trainer.placement.colocate_all": true, "trainer.use_sample_packing": false}'
```

Cookbook
```bash
 TINKER_API_KEY=tml-dummy uv run --with tinker --with datasets --with torch python -m \
    tinker_cookbook.recipes.vlm_classifier.train  \
    base_url=http://localhost:8000 \
    model_name="Qwen/Qwen3-VL-4B-Instruct" \
    dataset=caltech101 
```

Train nll:
train_nll

Val nll:
val_nll

Val accuracy:
accuracy

---

<a href="https://app.devin.ai/review/novasky-ai/skyrl/pull/1496" target="_blank">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-open-in-devin-review-dark.svg?v=1">
    Open with Devin
  </picture>
</a>

### CODE REVIEW

File: skyrl/backends/skyrl_train_backend.py [hunk 1]
```diff
@@ -510,8 +540,8 @@ def forward_backward(
         if not prepared_batch.all_model_inputs:
             return {}
 
-        self._sleep_inference_engines()
         batch = self._to_training_batch(prepared_batch)
+        self._sleep_inference_engines()
```
[the remark this answers is not in the document]
Comment (nithinvc) [reply skyrl/backends/skyrl_train_backend.py:541]: I tested this. vLLM doesn't need the actual vision tower or model loaded in order to render. One advantage (and justification) of the disaggregated render and generate at vllm was so that render can be run on cpu workers separately.

File: skyrl/backends/skyrl_train_backend.py [hunk 2]
```diff
@@ -425,6 +434,21 @@ def _to_training_batch(self, prepared_batch: types.PreparedModelPassBatch) -> Tr
...
+        pv_tensors = [
+            r.multi_modal_kwargs["pixel_values"]
+            for r in rendered_inputs
+            if r.multi_modal_kwargs is not None and r.multi_modal_kwargs.get("pixel_values") is not None
+        ]
+        thw_tensors = [
+            r.multi_modal_kwargs["image_grid_thw"]
+            for r in rendered_inputs
+            if r.multi_modal_kwargs is not None and r.multi_modal_kwargs.get("image_grid_thw") is not None
+        ]
```
[the remark this answers is not in the document]
Comment (nithinvc) [reply skyrl/backends/skyrl_train_backend.py]: fixed

File: skyrl/backends/skyrl_train_backend.py [hunk 3]
```diff
@@ -425,6 +434,21 @@ def _to_training_batch(self, prepared_batch: types.PreparedModelPassBatch) -> Tr
...
+        if pv_tensors:
+            batch_dict["pixel_values"] = TensorList(pv_tensors)
+        if thw_tensors:
+            batch_dict["image_grid_thw"] = TensorList(thw_tensors)
```
[the remark this answers is not in the document]
Comment (nithinvc) [reply skyrl/backends/skyrl_train_backend.py]: fixed

File: skyrl/backends/skyrl_train_backend.py [hunk 4]
```diff
@@ -374,8 +376,15 @@ def _to_training_batch(self, prepared_batch: types.PreparedModelPassBatch) -> Tr
         if not prepared_batch.all_model_inputs:
             return TrainingInputBatch({})
 
-        # Extract token IDs from ModelInput chunks
-        all_input_ids = [r.prompt_ids for r in render_model_input(prepared_batch.all_model_inputs)]
+        if _SKYRL_USE_NEW_INFERENCE:
+            if self._renderer is None:
+                self._ensure_inference_engines()
+                self._renderer = VLLMRenderer(self._inference_engine_client, self._cfg.trainer.policy.model.path)
+            rendered_inputs = asyncio.run(self._renderer(prepared_batch.all_model_inputs))
```
[the remark this answers is not in the document]
Comment (nithinvc) [reply skyrl/backends/skyrl_train_backend.py:383]: This is good, and in general I think the entire backend should move to a single persistent event loop. However, the scope of this PR is only on correctness for now unless requested

**Review by pcmoritz — APPROVED** [review]
